File: 정수빈/LeetCode/medium/binary-tree-level-order-traversal.py
# https://leetcode.com/problems/binary-tree-level-order-traversal
# 시간: 4분 40초
# 난이도: medium
# 설명: queue를 이용한 bfs

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root:
            return []
        q = collections.deque()
        visited = set()
        q.append((root, 0))
        depth = {}

        while q:
            node, level = q.popleft()
            visited.add(node)
            if level not in depth:
                depth[level] = []
            depth[level].append(node.val)

            if node.left and node.left not in visited:
                q.append((node.left, level+1) )
            if node.right and node.right not in visited:
                q.append((node.right, level+1))
        
        ans = []
        for key in sorted(depth.keys()):
            ans.append(depth[key])
        return ans

File: 장세훈/LeetCode/medium/BinaryTreeLevelOrderTraversal.java
class BinaryTreeLevelOrderTraversal {

    /*
        https://leetcode.com/problems/binary-tree-level-order-traversal/
     */

    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> result = new ArrayList<>();
        if(root ==null) return result;

        Deque<TreeNode> queue = new ArrayDeque<>();
        queue.add(root);

        while(!queue.isEmpty()) {
            int size = queue.size();
            List<Integer> list = new ArrayList<>();

            for (int i = 0; i < size; i++) {
                TreeNode node = queue.pollFirst();
                list.add(node.val);
                if(node.left != null) queue.addLast(node.left);
                if(node.right != null) queue.addLast(node.right);
            }
            result.add(list);
        }

        return result;
    }
}